### ISSUE TITLE
nightly aarch64: broken in debug mode

### ISSUE THREAD

**Issue by SWW13** [rust-lang/rust#48673]
The code below does not run correctly under _aarch64_ on the _nightly_ compiler in _debug_ mode.
Every other combination **works** (only tried _nightly_ and _stable_ releases).

I minimized the sample as far as I could and verified it on real hardware.
Sorry for the title, I have no clue how to name this issue.

I tried this code:

```rust
#[derive(Debug, PartialEq)]
pub struct B {
    pub b1: u16,
    pub b2: u8,
}

impl B {
    fn new(b1: u16, b2: u8) -> B {
        B { b1, b2 }
    }
}

#[test]
fn welp() {
    assert_eq!(B::new(0xDEAD, 0xFE), B { b1: 0xDEAD, b2: 0xFE });
}
```

I expected to see this happen:
```
$ cargo +nightly test
[...]
test welp ... ok
```

Instead, this happened:

(qemu-aarch64)
```
$ cargo +nightly test
[...]
test welp ... thread 'welp' panicked at 'assertion failed: `(left == right)`
  left: `B { b1: 0, b2: 0 }`,
 right: `B { b1: 57005, b2: 254 }`', src/lib.rs:15:5
```

(aarch64 cpu)
```
  left: `B { b1: 127, b2: 0 }`,
 right: `B { b1: 57005, b2: 254 }`', src/lib.rs:15:5
```

## Meta

You can reproduce this with qemu-user (aarch64) with the following config:
```
$ cat .cargo/config
[target.aarch64-unknown-linux-gnu]
linker = "aarch64-linux-gnu-gcc"
runner = "qemu-aarch64"
$ LD_LIBRARY_PATH="/usr/aarch64-linux-gnu/lib:$LD_LIBRARY_PATH" cargo +nightly test --target aarch64-unknown-linux-gnu
```

CPU:
```
$ lscpu
Architecture:          aarch64
Byte Order:            Little Endian
[...]
$ cat /proc/cpuinfo
[...]
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4
[...]
```

`rustc --version --verbose`:
```
rustc 1.26.0-nightly (3eeb5a665 2018-03-01)
binary: rustc
commit-hash: 3eeb5a665e313c5b281820099e04d4c6c8188b46
commit-date: 2018-03-01
host: x86_64-unknown-linux-gnu
release: 1.26.0-nightly
LLVM version: 6.0
```

---

**Edit:** smaller example

**Comment by parched** [comment]
I narrowed this down a bit to being caused by the LLVM 6 upgrade. I'll try create a minimal LLVM IR repro.

**Comment by parched** [comment]
Looks like [LLVM 6 enables GlobalISel](http://prereleases.llvm.org/6.0.0/rc2/docs/ReleaseNotes.html#changes-to-the-aarch64-target) so I tried with `RUSTFLAGS="-C llvm-args=-fast-isel"` and the problem goes away.

**Comment by parched** [comment]
There's some changes to the AArch64InstructionSelector in https://github.com/rust-lang/rust/pull/48782 for inserting/extracting 16bit types so hopefully that will fix it.

**Comment by nikomatsakis** [comment]
triage: P-high

Check-in from compiler team meeting: thanks for the investigative work! We'll keep an eye on this and see what happens when #48782 lands.

**Comment by parched** [comment]
I tested this on the latest nightly and it's fixed, so it seems the upgrade to released LLVM 6.0 indeed did the trick.

**Comment by nikomatsakis** [comment]
OK, then I'm going to close the issue! 🎷
